FREYA-1640: New SLU wastewater dashboard

senthil10 opened the pull request:
### 📋 Summary
<!-- Briefly describe WHAT and WHY of this PR -->
This PR adds completely new SLU wastewater dashboard combining all active wastewater surveillance under single dashboard. This new layout was proposed by the researchers.
 
### 🛠️ Changes Made
<!-- One-liners or bullet points -->
- a (temp) data sync page is added to sync data from blob server to DB
- added templates related to this dashboard in `templates/dashboards/slu_ww`
- added related views in `views/slu_ww.py`
- added visualisation script in `visualisation/slu_ww.py`

### 🔍 Notes for Reviewers
<!-- Tips, edge cases, or test instructions -->
The rendered content is almost final (maybe except some styling and minor tweaks), how it is rendered is implemented with respect to the MVP. After deciding on the dashboards architecture in next phase, relevant modifications will be made.

To test this dashboard locally, kindly to the following steps.

1) If not done already, run django migrate to create the dashboard data table.
2) Log into to django admin panel (after login nothing need to be done)
3) Then go to `localhost:8000/dashboards/slu-wastewater/data-sync` and press `Sync SLU data` button
4) After this no special steps, you can go to the urls added in this PR and review the pages.

### ✅ Checklist
- [x] PR title follows the pattern: `FREYA-XXXX: Clear and short description`
- [x] Jira / Github issue is linked in description
- [x] Assignee is selected
- [x] Code and content adhere to [conventions](https://scilifelab.atlassian.net/wiki/spaces/CDP2/pages/1909424133/Guidelines+for+the+portal+content)
- [x] Automated checks pass
- [x] Reviewer is selected when the PR is marked as ready for review

### 🔗 Jira Issue
<!-- Example: FREYA-914 -->
Closes: [FREYA-1640](https://scilifelab.atlassian.net/browse/FREYA-1640)

Ziip-dev reviewed (changes requested):
Given #60 , maybe we can already leverage it here so we have a solid baseline implementation that can serve as an example for other `htmx` stuff to come?

Ziip-dev commented on pages/dashboards/views/slu_ww.py:
```diff
@@ -0,0 +1,155 @@
+from ..visualisation.utils import plot_html_from_json
+
+
+class SluWasterWater(View):
```
```suggestion
class SluWasteWater(View):
```

typo :)

senthil10 replied:
Gahh!! I am spoiled by VS code auto addition and made the same typo in multiple files 😆 

Fixed it now, thank you for spotting.

Ziip-dev commented on pages/dashboards/views/__init__.py:
```diff
@@ -21,6 +22,8 @@
     "PostCovid",
     "SerologyStatistics",
+    "SLUsync",
+    "SluWasterWater",
```
```suggestion
    "SluWasteWater",
```

Ziip-dev commented on pages/dashboards/urls.py:
```diff
@@ -1,11 +1,14 @@
     PostCovid,
     SerologyStatistics,
+    SLUsync,
+    SluWasterWater,
```
```suggestion
    SluWasteWater,
```

Ziip-dev commented on pages/dashboards/urls.py:
```diff
@@ -58,3 +61,22 @@
 ]
+
+# SLU wastewater URLs
+for page in SluWasterWater.pages:
```
```suggestion
for page in SluWasteWater.pages:
```

Ziip-dev commented on pages/dashboards/views/slu_ww.py:
```diff
@@ -0,0 +1,155 @@
+from core.settings.base import env  # noqa: E402, F401
+
+
+class SLUsync(LoginRequiredMixin, View):
```
Then we can use `trigger_client_event` and `HX-Trigger` to simplify the data sync by moving the logic server-side, something like:

```python
from django_htmx.http import trigger_client_event

class SLUsync(LoginRequiredMixin, View):
    def post(self, request, *args, **kwargs):
        ...
        response = HttpResponseClientRefresh()  # or HttpResponse("OK")
        trigger_client_event(response, "slu-sync:done", {"
```

And the template can listen for `htmx:afterOnLoad` (or your `slu-sync:done` if you prefer) to avoid custom button toggling JS:
```html
<button hx-post="{% url 'dashboards:slu_data_sync' %}" hx-trigger="click" hx-indicator=".spinner">
    Sync SLU data
</button>
<div id="sync-status"></div>
<script>
    document.body.addEventListener('slu-sync:done', (event) => {
        document.querySelector('#sync-status').textContent =
            `Synced ${event.detail.rows} records`;
    });
</script>
```

Ziip-dev reviewed (changes requested):
Sorry I didn't see your answer while I was doing this (and I felt bad about giving you extra work on this one 😅) so here are some examples and guidelines if that can help:

senthil10 commented on pages/dashboards/templates/dashboards/slu_ww/base.html:
> * Load `htmx` once in `core/templates/base.html` (to discuss, but if we end up using `htmx` capabilities more and more throughout the pages I think it's worth it)

I thought about this too, not just for this but for `plotly js` too. Like having it in the main `base.html` but with a condition check like

```
{% if load_htmx %} {% htmx_script %} {% endif %}
{% if load_plotlyjs %} {% include "load_plotlyjs.html" %} {% endif %}
```

Then the above variables can be set in the respective views to include these libraries.

> * Use `request.htmx.boosted` in the views to `hx-boost` the tab navigation container in `base.html` so tab clicks fetch only the content area:
> 
> ```
> <div hx-boost="true" hx-target="main" hx-push-url="true">
> ```

Hmm I am not a fan of using `htmx` to navigate through the pages considering the overhead (just loading some libraries which are mostly cached by browsers) we would potentially save. I also prefer not to give an illusion of a single-page-application while it is not. I would prefer to use `htmx` only to update parts of the page and not the whole main content unless it is necessary 🙂

Ziip-dev replied:
```
{% if load_htmx %} {% htmx_script %} {% endif %}
{% if load_plotlyjs %} {% include "load_plotlyjs.html" %} {% endif %}
```
That's cool! Then we would be able to load it like `context["load_htmx"] = True` ?

Well the dashboard is a part of the page technically ^^ But I get your point, I'm curious why though, if you don't mind explaining, because I would genuinely be like '_user click on a button so let's just reload the dashboard area and call it a day without having different endpoints for different sections of the dashboard_'

senthil10 replied:
> ```
> {% if load_htmx %} {% htmx_script %} {% endif %}
> {% if load_plotlyjs %} {% include "load_plotlyjs.html" %} {% endif %}
> ```
> 
> That's cool! Then we would be able to load it like `context["load_htmx"] = True` ?

Exactly, thats what I was thinking. 

> Well the dashboard is a part of the page technically ^^ 

Yes, but each tab is its own page (i.e. everything except header menu bar and footer changes). Which is basically the whole page.

> But I get your point, I'm curious why though, if you don't mind explaining, because I would genuinely be like '_user click on a button so let's just reload the dashboard area and call it a day without having different endpoints for different sections of the dashboard_'

As mentioned above, each tab is its own page and some of those pages/tabs (Overview and the pathogens pages/tabs) have graphs.

Currently, we use `htmx` to only update these graphs based and selected inputs/filters and use the native link (`a`) navigation to change tabs. So we know for a given url if it is a normal request then we need load the whole page and it is a `htmx` request we need to update the plot.

If we use `htmx` to also navigate the tabs, that would mean a `htmx` request could either load the whole page or just the plot. That means the view should check where the request was trigged from for the `htmx` request. I feel like we wont be gaining much, but just making the view little complex than needed. Also in future if something changes, it could also get more complex.

Thus in general, I would like to avoid loading the whole page using the `htmx` as much as possible. To `htmx` is to load parts of the page not the whole page 🙂

Ziip-dev replied:
Ok, I wouldn't mind using those separate pages (the tabs) as template partials and load them in place but that's just me :)
Otherwise, fair point! I don't see any strong argument for or against it so it's just as good

senthil10 commented on pages/dashboards/views/slu_ww.py:
This would mean that the check is done on server side right ? which means the request is made ? But I have the check in client side to prevent making the request if not any years or sites are selected.

Also why to do page refresh (via `HttpResponseClientRedirect `), that means all other selected filters would be lost ?

senthil10 replied:
> The client check stays, I was just mentioning `HttpResponseClientRedirect` as a cheap server fallback so we never send empty response: e.g. to handle a request with invalid filters (if even possible).
> 
> But it can also be simple htmx tests on some get parameters followed by a usual render template 🤔

Since the filters values are created from the raw data file, there should never be invalid filters/parameters. Even on exception case, I think we should give the appropriate response (message) about the selected parameters rather than a redirect or reload :)

Ziip-dev commented:
I'll go through it today! In the meantime, can you rebase on `main`?

senthil10 commented:
> I think HTMX requests crash when no dashboard data. But i do not completely understand this workflow.

Yeah, my bad. I implemented preliminary lazy loading  after I implemented the data check handling in the view, thus din't cover the scenario for lazy loading while the data doesn't exist. Will do a implementation now.

> * **When:** A user opens the SLU wastewater Overview (or any virus tab) **before** any data has been synced, or when the dashboard row doesn’t exist yet. The initial page load can succeed (you set `no_recent_data_info` etc.), but as soon as HTMX fires (e.g. `hx-trigger="load once"` in `overview.html`), the view is called with `request.htmx=True` and `raw_data=None`.

The view is actually not called at all because `htmx` doesn't make a request. There is pre-check that atleast one year is selected before `htmx` sends a request, but since there is no inputs all (meaning no year) it fails and say "Select atleast one year" 😅  

> Morevoer there are places like POST requests that are missing error handling and exception handling which i think can be a big task and can be avoided for now as we decided to revisit dashboards and revise workflows for this.

Yes, the `POST` request is temp, thus I made a very minimal working `class`. This class will be removed fully.

senthil10 commented on pages/dashboards/views/slu_ww.py:
```diff
@@ -0,0 +1,185 @@
+
+    # Dashboard related values
+    dashboard = "slu_wastewater"
+    data_url = "https://blobserver.dc.scilifelab.se/blob/slu_test_data.csv"
```
Yes, I implemented like that previously, but since this is temporary and be removed later @Ziip-dev suggested we can just hardcode it. I too agree with that, so hardcoded it here, but I should have updated the PR description accordingly.

misterabdullahAziz reviewed (approved):
Thanks for the updates. 
We’re fine with keeping data_url hardcoded and the sync POST minimal for this temp class.
Approved for merge.
